### DISCUSSION TITLE
[Student import] While upserting organization refs, how do we handle organizations which do not exist?

### DISCUSSION THREAD

**Discussion by netsensei** [ugent-library/old-people-service#38] [General]
In `cmd/students_import.go` we find this code block:

```go
for _, oRef := range np.Organization {
	realOrg, err := repo.GetOrganizationByOtherId(ctx, "ugent_id", oRef.Id)
	if errors.Is(err, models.ErrNotFound) {
		continue
	} else if err != nil {
		return err
	}
        newOrgRef := models.NewOrganizationRef(realOrg.Id)
        newOrgRef.From = oRef.From
        newOrgRef.Until = oRef.Until
        newOrgRefs = append(newOrgRefs, newOrgRef)
}
```

This code block will iterate over the `Organization` property on the preson - as is provided via LDAP - and then look for already existing organizations in the database. If it finds an organization, it will add a reference to the `Organization` property of the person object that will be saved to the database. 

However, if it can't find the organization, it just throws an `ErrNotFound` error and moves on to the next item in the `Organization` slice. **As a result, we risk losing references to potentially relevant organizations.**

Either:

* A stub organization is automatically created in the database.
* A clear error is thrown in the output, allowing us to pick up organizations that don't exist, adding them manually.

**Comment by nics** [comment]
I would just add a memory only dummy org to the slice of orgs and log a warning, that's what biblio-backoffice does now. It's not necessarily an exception if the org don't exist in the db yet, could just be a timing issue.

**Comment by nicolasfranck** [comment]
this command is supposed to run every night. So what was not found one night might get found another night.
So I see no reason to add a dummy organization. Would be better though to log an warning.

@nics I cannot add a memory only reference. All organizations are added through an intermediary table that links organizations to users. We removed `other_organization_id`, remember? Or I just add a real organization with stub data (e.g. organization "BIB" with name_dut "BIB" and name_eng "BIB"), and wait for the real data to overwrite it.

**Comment by netsensei** [comment]
@nics @nicolasfranck 

Okay, So there are three possible approaches to this issue:

**1/ Don't store the organization until the next day**

* This is the current logic, treat all missing data as a timing issue
* If an organization can't be found, the reference is dropped on the person side / not stored.
* A warning message is logged to indicate that an organization is missing.

This promotes eventual consistency.

Downside: Some data is missing and may only become available the day after. Users might potentially notice this in e.g. contributors suggest (not all affiliations listed), or when creating a new biblio record where the affiliation of the creator is automagically filled out in the "department" field (it won't be filled out due to missing data).

Upside: easy to implement.

**2/ Store a reference in the `OrganizationPerson` table**

* The `OrganizationPerson` table is expanded to have these fields:
  * organization_id
  * person_id
  * organization_public_id
  * from
  * until
* If an organization doesn't exist, do make a record in the `OrganizationPerson` table
  * Fill out the  `organization_public_id` with the id of the organization
  * make the `organization_id` nil. 
* Log with a warning message
* On the next pass, check if the person already has a reference to a non-existing organization:
  * create the organization
  * add the id to `organization_id` 


Downside: the OrganizationPerson table will contain references to non-existing organizations, that might never exist if the data is inconsistent. (note: we do work with external data which we do not control, nor shouldn't control)

Upside: This allows us to at least list all organizations per person, even those that are (still) missing. This also allows us to mitigate issues with organizations that aren't related to timing e.g. dirty data that we might need. This keeps the organization table itself from dummy organizations that won't ever be cleaned up.

**3/ Create a dummy record in the organization table**

* Always create an organization the organizations table. Create a dummy organization if no organization exists.

Downside: Downside: the Organization table will contain references to non-existing organizations, that might never exist if the data is inconsistent. (note: we do work with external data which we do not control, nor shouldn't control)

Upside:  This allows us to at least list all organizations per person, even those that are (still) missing. This also allows us to mitigate issues with organizations that aren't related to timing e.g. dirty data that we might need. This keeps the `OrganizationPerson` table itself from dummy organizations that won't ever be cleaned up.

Let's vote / pick an option and go with it.

**Reply by nicolasfranck** [reply]
I would vote for option 3. If the issue doesn't get cleaned, it is at least shown to everyone. In the past I worked with things like `other_organization_id` that would store id's of organizations that did not arrive yet, and then I upgraded those to real organizations (with a link in organization_person) when they came in. Like in option 2, I showed only the validated organizations to end users, and hided the others. But that generated unnecessary complication. Because if they never showed up, only a database user would be able to see those.

Anyway, we should "trust" incoming user data. If they make wrong statements, that is another issue.
